delira/data_loading/augmenter.py

```diff
@@ -0,0 +1,520 @@
+
+    def _shutdown_processes(self):
+        """
+        Shuts down the processes and resets all related flags and counters
+        """
+
+        # create copy to avoid modifying the list we iterate over
+        worker = list(
+            zip(self._data_pipes, self._index_pipes, self._processes))
+        # for each process:
+        for _data_conn, _index_conn, _process in worker:
+
+            # send None to worker
+            _index_conn.send(None)
+            # Close Process and wait for its termination
+            _process.join()
+            if sys.version_info >= (3, 7):
+                _process.close()
+            else:
+                _process.terminate()
+
+            # close connections
+            _index_conn.close()
+            _data_conn.close()
+
+            # pop corresponing pipes, counters and processes from lists
+            self._data_pipes.pop()
+            self._data_queued.pop()
+            self._index_pipes.pop()
+            self._processes.pop()
+
+        # reset running process flag and counters
+        self._processes_running = False
+        self._data_pipe_counter = 0
+        self._index_pipe_counter = 0
+
+    @property
+    def _next_index_pipe(self):
+        """
+        Property implementing switch to next index pipe
+        """
+        ctr = self._index_pipe_counter
+        new_ctr = (self._index_pipe_counter + 1) % self._num_processes
+        self._index_pipe_counter = new_ctr
+
+        return ctr
+
+    @property
+    def _next_data_pipe(self):
+        """
+        Property implementing switch to next data pipe
+        """
+        ctr = self._data_pipe_counter
+        new_ctr = (self._data_pipe_counter + 1) % self._num_processes
+        self._data_pipe_counter = new_ctr
+
+        return ctr
+
+    def _enqueue_indices(self, sample_idxs):
+        """
+        Enqueues a set of indices to workers while iterating over workers in
+        cyclic way
+        Parameters
+        ----------
+        sample_idxs : list
+            the indices to enqueue to the workers
+        """
+
+        # iterating over all batch indices
+        for idxs in sample_idxs:
+            # switch to next counter
+            index_pipe_ctr = self._next_index_pipe
+            # increase number of queued batches for current worker
+            self._data_queued[index_pipe_ctr] += 1
+            # enqueue indices to worker
+            self._index_pipes[index_pipe_ctr].send(idxs)
+
+    def _receive_data(self):
+        """
+        Receives data from worker
+        """
+        # switching to next worker
+        _data_pipe = self._next_data_pipe
+
+        # receive data from worker
+        data = self._data_pipes[_data_pipe].recv()
+        # decrease number of enqueued batches for current worker
+        self._data_queued[_data_pipe] -= 1
+
+        return data
+
+    def __iter__(self):
+        # start processes
+        self._start_processes()
+
+        # create sampler_old iterator
```

Comment:
Nit: Comment might be unnecessary, 
What is `sampler_old` iterator and where is the difference to a "new" sampler?

Reply:
This is not up to date anymore. This was part of the intermediate state of this PR, simply forgot to change this